packages/alfa-rules/src/common/diagnostic/contrast.ts

```diff
@@ -7,22 +7,24 @@ import * as json from "@siteimprove/alfa-json";
 import { Hash, Hashable } from "@siteimprove/alfa-hash";
 import { Comparable, Comparison } from "@siteimprove/alfa-comparable";
 
+type Name = ["container", "link"] | ["foreground", "background"];
+
 export class Contrast extends Diagnostic {
```

Comment:
Let's add the `<N extends Name>` parameter on `Contrast` (so we can forward it to the `Pairing` inside).

To make `Color` cleaner, let's also add a `type FirstColor<N extends Name> = N[0]` (and `SecondColor`), the `Color` type parameter would then extend `FirstColor<Name> | SecondColor<Name>` rather than `Name` and the `N[0] | N[1]` ambiguity can be removed there. The `_color1`, `_color2` private variables (and associated getters, …) in `Pairing` can then be typed directly typed `FirstColor<N>`, … so that the type system is aware of the colors' names all along.

Doing it that way (rather than directly using `N[0]`) also makes it easy to change the representation of `Name` if we decide it's not the best, we only need to adapt `FirstColor` and the rest flows.